File: app/ios/App/App/AchievementsPlugin.swift
import Capacitor
import GameKit

/* THE STORE'S ACHIEVEMENTS ON iOS: Game Center. Two calls, and
   js/27-achievements.js is the only caller:

     signIn()        -> {signedIn}  sign the player in to Game Center
     unlock({id})    -> {ok}        report one achievement at 100%

   It lives in the app rather than in an npm package because no maintained
   plugin fits Capacitor 7, and what is needed is two calls. A plugin in the
   app is not found by `cap sync`, so AppViewController registers it, and
   Main.storyboard names AppViewController instead of Capacitor's own.

   Game Center also needs its entitlement (App.entitlements) and the
   achievements created in App Store Connect under the ids in ACH_IDS.ios. */
@objc(AchievementsPlugin)
public class AchievementsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AchievementsPlugin"
    public let jsName = "Achievements"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "signIn", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unlock", returnType: CAPPluginReturnPromise)
    ]

    /* Calls waiting for Game Center's answer. The handler can only be set
       once per launch - setting it IS the sign-in - so a second signIn()
       joins the first instead of starting another. */
    private var waiting: [CAPPluginCall] = []
    private var asked = false
    private var answered = false

    @objc func signIn(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            let me = GKLocalPlayer.local
            if me.isAuthenticated || (self.asked && self.answered) {
                call.resolve(["signedIn": me.isAuthenticated])
                return
            }
            self.waiting.append(call)
            if self.asked { return }
            self.asked = true
            /* Apple's pattern: a player already signed in on this phone is
               signed in silently with a "Welcome back" banner; one who is not
               is handed a sign-in sheet, which we present. Cancelling it is
               an answer, and Game Center itself stops offering the sheet
               after a few cancels. */
            me.authenticateHandler = { [weak self] vc, _ in
                guard let self = self else { return }
                if let vc = vc {
                    self.bridge?.viewController?.present(vc, animated: true)
                    return
                }
                self.answered = true
                let ok = GKLocalPlayer.local.isAuthenticated
                let calls = self.waiting
                self.waiting = []
                for c in calls { c.resolve(["signedIn": ok]) }
            }
        }
    }

    @objc func unlock(_ call: CAPPluginCall) {
        guard let id = call.getString("id"), !id.isEmpty,
              GKLocalPlayer.local.isAuthenticated else {
            call.resolve(["ok": false])
            return
        }
        let a = GKAchievement(identifier: id)
        a.percentComplete = 100
        a.showsCompletionBanner = true
        GKAchievement.report([a]) { error in
            call.resolve(["ok": error == nil])
        }
    }
}

File: app/android/app/src/main/java/com/nadazgames/imjustacube/AchievementsPlugin.java
package com.nadazgames.imjustacube;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.android.gms.games.PlayGames;
import com.google.android.gms.games.PlayGamesSdk;

/* THE STORE'S ACHIEVEMENTS ON ANDROID: Google Play Games Services v2.
   Two calls, and js/27-achievements.js is the only caller:

     signIn()        -> {signedIn}  is this player signed in to Play Games
     unlock({id})    -> {ok}        tell Play one achievement is unlocked

   It lives in the app rather than in an npm package because no maintained
   plugin fits Capacitor 7 (the ones that exist are 0.x and need 8), and what
   is needed is two calls. MainActivity registers it.

   NOT SET UP IS NOT AN ERROR. Until Play Console has made a Games Services
   project and its id is in res/values/strings.xml, `game_services_project_id`
   is empty, and then the SDK is never started and every call answers "no".
   That matters because the SDK's own start-up, with no id, is the "fatal
   developer error" crash - which is also why its automatic start-up
   (PlayGamesInitProvider) is removed in AndroidManifest.xml and started here
   by hand instead, only when there is an id to start it with. */
@CapacitorPlugin(name = "Achievements")
public class AchievementsPlugin extends Plugin {
    private boolean ready = false;

    @Override
    public void load() {
        if (!configured()) return;
        try {
            PlayGamesSdk.initialize(getContext());
            ready = true;
        } catch (Exception e) {
            ready = false;
        }
    }

    /* The manifest's APP_ID, read back: present, non-empty and a number.
       Bundle.get() is deprecated, and kept: the value is a String from the
       string resource, but an Integer if anybody types the number in. */
    @SuppressWarnings("deprecation")
    private boolean configured() {
        try {
            ApplicationInfo ai = getContext().getPackageManager().getApplicationInfo(
                getContext().getPackageName(), PackageManager.GET_META_DATA);
            Bundle md = ai.metaData;
            Object v = md == null ? null : md.get("com.google.android.gms.games.APP_ID");
            String s = v == null ? "" : String.valueOf(v).trim();
            return s.matches("[0-9]+") && !s.equals("0");
        } catch (Exception e) {
            return false;
        }
    }

    /* Play Games v2 signs the player in by itself at launch, so this only
       asks. It never puts up a sign-in screen: somebody who declined Play
       Games once is not asked again from inside a game about a cube. */
    @PluginMethod
    public void signIn(PluginCall call) {
        if (!ready || getActivity() == null) {
            JSObject r = new JSObject();
            r.put("signedIn", false);
            call.resolve(r);
            return;
        }
        PlayGames.getGamesSignInClient(getActivity()).isAuthenticated()
            .addOnCompleteListener(t -> {
                JSObject r = new JSObject();
                r.put("signedIn", t.isSuccessful() && t.getResult() != null
                                  && t.getResult().isAuthenticated());
                call.resolve(r);
            });
    }

    /* unlock() rather than unlockImmediate(): Play queues it and sends it
       when there is a connection, so an achievement earned on a train is
       not lost. Answering ok as soon as it is queued is therefore honest. */
    @PluginMethod
    public void unlock(PluginCall call) {
        String id = call.getString("id");
        JSObject r = new JSObject();
        if (!ready || getActivity() == null || id == null || id.isEmpty()) {
            r.put("ok", false);
            call.resolve(r);
            return;
        }
        try {
            PlayGames.getAchievementsClient(getActivity()).unlock(id);
            r.put("ok", true);
        } catch (Exception e) {
            r.put("ok", false);
        }
        call.resolve(r);
    }
}
